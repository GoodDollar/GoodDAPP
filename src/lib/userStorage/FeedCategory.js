// @flow
import { t } from '@lingui/macro'

export type FeedCategory = null | 'news' | 'tx'

export class FeedCategories {
  static All = null

  static News = 'news'

  static Transactions = 'tx'

  static get all() {
    const { All, Transactions, News } = this

    return [All, Transactions, News]
  }

  static label(category) {
    const { News, Transactions } = this

    switch (category) {
      case News:
        return t`News`
      case Transactions:
        return t`Transactions`
      default:
        return t`All`
    }
  }
}
