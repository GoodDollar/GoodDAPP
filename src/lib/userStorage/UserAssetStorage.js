// @flow
import { assign, pick } from 'lodash'

import IPFS from '../ipfs/IpfsStorage'
import { ThreadDB } from '../textile/ThreadDB'

export class UserAssetStorage {
  storage: any

  db: ThreadDB

  localCache: { [string]: { binary: boolean, dataUrl: string } } = {}

  static factory(db: ThreadDB): UserAssetStorage {
    return new UserAssetStorage(IPFS, db)
  }

  constructor(storage: any, db: ThreadDB) {
    assign(this, { storage, db })
  }

  // eslint-disable-next-line require-await
  async store(dataUrl: string): Promise<string> {
    const cid = await this.storage.store(dataUrl)

    await this._writeCache(cid, { dataUrl, binary: true })
    return cid
  }

  // eslint-disable-next-line require-await
  async load(cid: string, withMetadata: boolean = false): Promise<string | { binary: boolean, dataUrl: string }> {
    let avatar = await this._readCache(cid)

    if (!avatar) {
      avatar = await this.storage.load(cid, true)
      await this._writeCache(cid, avatar)
    }

    return withMetadata ? avatar : avatar.dataUrl
  }

  // async as invokes threaddb method to remove asset from the cache collection
  async clearCache(cid: string): Promise<void> {
    await this.db.Assets.delete(cid)
    delete this.localCache[cid]
  }

  async _readCache(cid: string): Promise<string | null> {
    const { localCache } = this
    let data = localCache[cid]

    if (!data) {
      data = await this.db.Assets.findById(cid)

      if (data) {
        localCache[cid] = pick(data, 'dataUrl', 'binary')
      }
    }

    return data
  }

  async _writeCache(cid: string, data: { dataUrl: string, binary: boolean }): Promise<void> {
    const { dataUrl, binary = true } = data
    const cachedData = { dataUrl, binary }

    await this.db.Assets.save({ _id: cid, ...cachedData })
    this.localCache[cid] = cachedData
  }
}

export default UserAssetStorage.factory
