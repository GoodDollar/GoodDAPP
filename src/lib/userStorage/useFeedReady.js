import { useCallback } from 'react'
import usePropsRefs from '../hooks/usePropsRefs'
import { useUserStorage } from '../wallet/GoodWalletProvider'

export const onFeedReady = async userStorage => {
  const { feedStorage, initializedRegistered } = userStorage

  if (!initializedRegistered) {
    await feedStorage.ready
  }
}

const useFeedReady = () => {
  const userStorage = useUserStorage()
  const { initializedRegistered } = userStorage
  const [getStorage] = usePropsRefs([userStorage])

  const onReady = useCallback(async () => {
    await onFeedReady(getStorage())
  }, [getStorage])

  return [initializedRegistered, onReady]
}

export default useFeedReady
