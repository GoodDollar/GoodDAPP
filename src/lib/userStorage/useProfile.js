import { pick } from 'lodash'
import { useCallback, useEffect, useMemo, useState } from 'react'

import { useUserStorage } from '../wallet/GoodWalletProvider'

const defaultPublicFields = ['fullName', 'smallAvatar']

const useProfileHook = (fields, allowRefresh = false, display = false) => {
  const [profile, setProfile] = useState({})
  const userStorage = useUserStorage()

  const getProfile = useCallback(
    (fields, display) => {
      // the function checks are for when in FV-Standalone flow
      // when redirect/reload within flow it cam lose the context if it is fvFlow or not since it is using
      // some shared components with the regular flow
      if (!userStorage || !userStorage.getDisplayProfile) {
        return {}
      }

      const profile = display ? userStorage.getDisplayProfile() : userStorage.getPrivateProfile()

      return fields ? pick(profile, fields) : profile
    },
    [userStorage],
  )

  const refreshProfile = useCallback(() => setProfile(getProfile(fields, display)), [
    fields,
    display,
    setProfile,
    getProfile,
  ])

  // auto refresh provide each time fields and private changes
  // also initializes profile on mount
  useEffect(() => void refreshProfile(), [refreshProfile])

  return useMemo(() => (allowRefresh ? [profile, refreshProfile] : profile), [profile, refreshProfile, allowRefresh])
}

const useProfile = (allowRefresh = false, fields = null) => useProfileHook(fields, allowRefresh)

export const usePublicProfile = (allowRefresh = false, fields = null) => useProfileHook(fields, allowRefresh, true)

export const usePublicProfileOf = (walletAddress, fields = defaultPublicFields) => {
  const [profile, setProfile] = useState(null)
  const userStorage = useUserStorage()

  useEffect(() => {
    userStorage
      .getPublicProfile('walletAddress', walletAddress)
      .then(rawProfile => setProfile(pick(rawProfile, fields)))
  }, [walletAddress, fields, setProfile])

  return profile
}

const getUserProperty = (userStorage, property, local = false) => {
  if (!userStorage) {
    return null
  }

  const { userProperties } = userStorage

  return local ? userProperties.getLocal(property) : userProperties.get(property)
}

export const useUserProperty = (property, local = false) => {
  const userStorage = useUserStorage()

  const [propertyValue, setPropertyValue] = useState(() => getUserProperty(userStorage, property, local))

  const updatePropertyValue = useCallback(
    newValue => {
      const { userProperties } = userStorage

      setPropertyValue(newValue)

      if (local) {
        userProperties.setLocal(property, newValue)
        return
      }

      userProperties.safeSet(property, newValue)
    },
    [setPropertyValue, userStorage, property, local],
  )

  useEffect(() => {
    const { userProperties } = userStorage

    setPropertyValue(getUserProperty(userStorage, property, local))

    if (!local) {
      userProperties.on(property, setPropertyValue)
    }

    return () => {
      if (local) {
        return
      }

      userProperties.off(property, setPropertyValue)
    }
  }, [property, userStorage, setPropertyValue, local])

  return [propertyValue, updatePropertyValue]
}

export const useLocalProperty = property => useUserProperty(property, true)

export default useProfile
