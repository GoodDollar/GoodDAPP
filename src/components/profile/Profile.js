// @flow
import React, { useCallback } from 'react'
import { Platform, View } from 'react-native'
import { createStackNavigator } from '../appNavigation/stackNavigation'
import { Section, Text, Wrapper } from '../common'
import UserAvatar from '../common/view/UserAvatar'
import { withStyles } from '../../lib/styles'
import { getDesignRelativeWidth } from '../../lib/utils/sizes'
import RoundIconButton from '../common/buttons/RoundIconButton'
import { usePublicProfile } from '../../lib/userStorage/useProfile'
import { theme } from '../theme/styles'
import EditAvatar from './EditAvatar'
import EditProfile from './EditProfile'
import ProfileDataTable from './ProfileDataTable'
import ProfilePrivacy from './ProfilePrivacy'
import ViewAvatar from './ViewOrUploadAvatar'
import VerifyEdit from './VerifyEdit'
import VerifyEditCode from './VerifyEditCode'

const avatarSize = getDesignRelativeWidth(136)

const ProfileWrapper = ({ screenProps, styles }) => {
  const profile = usePublicProfile()
  const { fullName } = profile

  const handleAvatarPress = useCallback(() => screenProps.push(`ViewAvatar`), [screenProps])

  const handlePrivacyPress = useCallback(() => screenProps.push(`ProfilePrivacy`), [screenProps])

  const handleEditProfilePress = useCallback(() => screenProps.push(`EditProfile`), [screenProps])

  return (
    <Wrapper>
      <Section.Row justifyContent="space-between" alignItems="flex-start" style={styles.userDataAndButtonsRow}>
        <RoundIconButton
          label={'Privacy'}
          iconName={'privacy'}
          iconSize={23}
          onPress={handlePrivacyPress}
          containerStyle={styles.iconLeft}
        />

        <RoundIconButton
          label={'Edit'}
          iconName={'edit'}
          iconSize={25}
          onPress={handleEditProfilePress}
          style={styles.iconRightContainer}
          containerStyle={styles.iconRight}
        />
      </Section.Row>
      <Section style={styles.section}>
        <View style={styles.emptySpace} />
        <ProfileDataTable profile={profile} showCustomFlag />
      </Section>
      <View style={styles.userDataWrapper}>
        <UserAvatar
          style={styles.userAvatar}
          profile={profile}
          onPress={handleAvatarPress}
          size={avatarSize}
          imageSize={avatarSize - 6}
          unknownStyle={styles.userAvatar}
        />
        {fullName ? (
          <Text fontSize={22} fontFamily={theme.fonts.slab} lineHeight={29} style={styles.userName}>
            {fullName}
          </Text>
        ) : null}
      </View>
    </Wrapper>
  )
}

ProfileWrapper.navigationOptions = {
  title: 'My Profile',
}

const getStylesFromProps = ({ theme }) => {
  const halfAvatarSize = avatarSize / 2
  return {
    emptySpace: {
      height: 75,
      width: '100%',
    },
    section: {
      flexGrow: 1,
      padding: theme.sizes.defaultDouble,
    },
    iconRightContainer: {
      transform: [{ rotateY: '180deg' }],
    },
    iconLeft: {
      position: 'absolute',
      left: getDesignRelativeWidth(20),
    },
    iconRight: {
      position: 'absolute',
      right: getDesignRelativeWidth(20),
    },
    userDataWrapper: {
      position: 'absolute',
      justifyContent: 'center',
      alignItems: 'center',
      alignSelf: 'center',
      zIndex: 1,
    },
    userAvatar: {
      borderWidth: 3,
      borderColor: theme.colors.white,
      justifyContent: 'center',
      alignItems: 'center',
      borderRadius: halfAvatarSize,
    },
    userDataAndButtonsRow: {
      display: 'flex',
      justifyContent: 'center',
      position: 'relative',
      zIndex: 1,
      height: avatarSize / 2,
    },
    userName: {
      marginTop: theme.sizes.default,
    },
  }
}

const Profile = withStyles(getStylesFromProps)(ProfileWrapper)

const commonRoutes = {
  Profile,
  EditProfile,
  ProfilePrivacy,
  ViewAvatar,
  VerifyEdit,
  VerifyEditCode,
}

const nativeOnlyRoutes = {}

const webOnlyRoutes = {
  EditAvatar,
}

const routes = {
  ...commonRoutes,
  ...Platform.select({
    web: webOnlyRoutes,
    default: nativeOnlyRoutes,
  }),
}

export default createStackNavigator(routes)
