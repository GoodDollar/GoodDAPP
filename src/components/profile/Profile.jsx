// @flow
import React, { useCallback } from 'react'
import { Text as NText, View } from 'react-native'
import { t } from '@lingui/macro'
import { GoodIdDetails } from '@gooddollar/good-design'

import { createStackNavigator } from '../appNavigation/stackNavigation'
import { Section, Text, Wrapper } from '../common'

import UserAvatar from '../common/view/UserAvatar'
import { withStyles } from '../../lib/styles'
import { getDesignRelativeWidth } from '../../lib/utils/sizes'
import RoundIconButton from '../common/buttons/RoundIconButton'
import { usePublicProfile } from '../../lib/userStorage/useProfile'
import { theme } from '../theme/styles'
import { useUserStorage, useWallet } from '../../lib/wallet/GoodWalletProvider'
import IdentifierRow from '../common/view/IdentifierRow'

import EditProfile from './EditProfile'
import ProfileDataTable from './ProfileDataTable'
import ViewAvatar from './ViewOrUploadAvatar'
import VerifyEdit from './VerifyEdit'
import VerifyEditCode from './VerifyEditCode'

const avatarSize = getDesignRelativeWidth(136)

const ProfileWrapper = ({ screenProps, styles }) => {
  const profile = usePublicProfile()
  const userStorage = useUserStorage()
  const goodWallet = useWallet()

  // const [faceRecordId, setRecordId] = useState()

  const logMethod = userStorage?.userProperties.get('logMethod')

  const { fullName } = profile

  const handleAvatarPress = useCallback(() => screenProps.push(`ViewAvatar`), [screenProps])

  const handleEditProfilePress = useCallback(() => screenProps.push(`EditProfile`), [screenProps])

  return (
    <Wrapper withMaxHeight={false}>
      <Section.Row justifyContent="space-between" alignItems="flex-start" style={styles.userDataAndButtonsRow}>
        <RoundIconButton
          label={'Edit'}
          iconName={'edit'}
          iconSize={25}
          onPress={handleEditProfilePress}
          style={styles.iconRightContainer}
          containerStyle={styles.iconRight}
        />
      </Section.Row>
      <Section style={styles.section}>
        <View style={styles.emptySpace} />
        <ProfileDataTable profile={profile} showCustomFlag />

        <Section grow justifyContent="flex-end" style={{ width: '100%', paddingLeft: 0, paddingRight: 0, margin: 0 }}>
          {logMethod && (
            <Section.Row>
              <IdentifierRow title="LoginM" text={logMethod} />
            </Section.Row>
          )}
        </Section>

        <View padding="0">
          <View padding="0">
            <NText
              style={{
                backgroundColor: '#00AEFF20',
                color: '#00AEFF',
                fontFamily: 'Montserrat',
                fontWeight: 700,
                fontSize: 24,
                paddingTop: 16,
                paddingBottom: 16,
                paddingLeft: 8,
                paddingRight: 8,
                width: '100%',
                textAlign: 'center',
                marginBottom: 24,
              }}
            >
              {' '}
              GoodID{' '}
            </NText>
            <GoodIdDetails account={goodWallet.account} isVerified={goodWallet.isVerified} />
          </View>
        </View>
      </Section>
      <View style={styles.userDataWrapper}>
        <UserAvatar
          style={styles.userAvatar}
          profile={profile}
          onPress={handleAvatarPress}
          size={avatarSize}
          imageSize={avatarSize - 6}
          unknownStyle={styles.userAvatar}
        />
        {fullName ? (
          <Text fontSize={22} fontFamily={theme.fonts.slab} lineHeight={29} style={styles.userName}>
            {fullName}
          </Text>
        ) : null}
      </View>
    </Wrapper>
  )
}

ProfileWrapper.navigationOptions = {
  title: t`My Profile`,
}

const getStylesFromProps = ({ theme }) => {
  const halfAvatarSize = avatarSize / 2

  return {
    emptySpace: {
      height: 115,
      width: '100%',
    },
    section: {
      flexGrow: 1,
      padding: theme.sizes.defaultDouble,
      paddingLeft: 8,
      paddingRight: 8,
      marginBottom: theme.paddings.bottomPadding,
    },
    iconRightContainer: {
      transform: [{ rotateY: '180deg' }],
    },
    iconLeft: {
      position: 'absolute',
      left: getDesignRelativeWidth(20),
    },
    iconRight: {
      position: 'absolute',
      right: getDesignRelativeWidth(20),
    },
    userDataWrapper: {
      position: 'absolute',
      justifyContent: 'center',
      alignItems: 'center',
      alignSelf: 'center',
      zIndex: 1,
    },
    userAvatar: {
      borderWidth: 3,
      borderColor: theme.colors.white,
      justifyContent: 'center',
      alignItems: 'center',
      borderRadius: halfAvatarSize,
    },
    userDataAndButtonsRow: {
      display: 'flex',
      justifyContent: 'center',
      position: 'relative',
      zIndex: 1,
      height: avatarSize / 2,
      margin: 0,
    },
    userName: {
      marginTop: theme.sizes.default,
    },
  }
}

const Profile = withStyles(getStylesFromProps)(ProfileWrapper)

const routes = {
  Profile,
  EditProfile,
  ViewAvatar,
  VerifyEdit,
  VerifyEditCode,
}

export default createStackNavigator(routes)
