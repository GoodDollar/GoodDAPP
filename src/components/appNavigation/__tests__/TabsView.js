import React from 'react'
import renderer from 'react-test-renderer'
import ImportedTabsView from '../TabsView'
import Dashboard from '../../dashboard/Dashboard'
import Profile from '../../profile/Profile'
import userStorage from '../../../lib/userStorage/UserStorage'
import { withThemeProvider } from '../../../__tests__/__util__/index'

// Note: test renderer must be required after react-native.
jest.setTimeout(30000)

describe('TabsView', () => {
  beforeAll(async () => {
    await userStorage.wallet.ready
    await userStorage.ready
  })

  const routes = {
    Dashboard: {
      screen: Dashboard,
    },
    Profile: {
      screen: Profile,
    },
  }

  const TabsView = withThemeProvider(ImportedTabsView)

  it('matches snapshot', async () => {
    let component

    // eslint-disable-next-line require-await
    await renderer.act(async () => (component = renderer.create(<TabsView routes={routes} />)))
    expect(component.toJSON()).toMatchSnapshot()
  })
})
