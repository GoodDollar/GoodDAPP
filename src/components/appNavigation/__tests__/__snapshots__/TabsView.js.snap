// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabsView matches snapshot 1`] = `
<div
  className="css-view-1dbjc4n r-flex-13awgt0 r-pointerEvents-12vffkv"
>
  <div
    className="css-view-1dbjc4n r-paddingBottom-97e31f r-paddingLeft-bv2aro r-paddingRight-hxflta r-paddingTop-10xqauy"
    style={
      Object {
        "backgroundColor": "rgba(0,175,255,1.00)",
        "boxShadow": "0px 3px 4px rgba(0,0,0,0.24)",
        "zIndex": 0,
      }
    }
  >
    <div
      className="css-view-1dbjc4n r-alignItems-1awozwy"
      style={
        Object {
          "WebkitBoxDirection": "normal",
          "WebkitBoxOrient": "horizontal",
          "WebkitBoxPack": "justify",
          "WebkitFlexDirection": "row",
          "WebkitJustifyContent": "space-between",
          "backgroundColor": "rgba(0,175,255,1.00)",
          "display": "flex",
          "flexDirection": "row",
          "height": "56px",
          "justifyContent": "space-between",
          "msFlexDirection": "row",
          "msFlexPack": "justify",
          "overflowX": "hidden",
          "overflowY": "hidden",
          "paddingLeft": "4px",
          "paddingRight": "4px",
        }
      }
    >
      <div
        className="css-view-1dbjc4n r-alignItems-1awozwy r-justifyContent-1777fci"
        style={
          Object {
            "WebkitBoxDirection": "normal",
            "WebkitBoxOrient": "horizontal",
            "WebkitFlexDirection": "row",
            "flexDirection": "row",
            "msFlexDirection": "row",
          }
        }
      >
        <div
          className="css-view-1dbjc4n r-alignItems-1awozwy r-cursor-1loqt21 r-flex-13awgt0 r-flexDirection-eqz5dr r-touchAction-1otgn73 r-transitionProperty-1i6wzkk r-userSelect-lrvibr"
          onClick={[Function]}
          onContextMenu={[Function]}
          onKeyDown={[Function]}
          style={
            Object {
              "WebkitTransitionDuration": "0s",
              "marginLeft": "8px",
              "transitionDuration": "0s",
            }
          }
          tabIndex="0"
        >
          <div
            className="css-view-1dbjc4n r-alignItems-1awozwy r-justifyContent-1777fci"
            style={
              Object {
                "backgroundColor": "rgba(0,0,0,0.00)",
              }
            }
          >
            <div
              className="css-text-901oao"
              dir="auto"
              style={
                Object {
                  "color": "rgba(255,255,255,1.00)",
                  "fontFamily": "gooddollar",
                  "fontSize": "16px",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                }
              }
            >
              
            </div>
          </div>
        </div>
        <div
          className="css-view-1dbjc4n"
          style={
            Object {
              "marginLeft": "4px",
              "marginRight": "4px",
            }
          }
        >
          <ellipse46.svg />
        </div>
        <div
          className="css-text-901oao r-color-jwli3a r-fontFamily-vnw8o6 r-fontSize-10x49cs r-fontWeight-b88u0q r-lineHeight-1f529hi r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-3twk1y"
          dir="auto"
        >
          DEVELOP
        </div>
      </div>
      <div
        className="css-view-1dbjc4n"
        style={
          Object {
            "marginLeft": "-50px",
            "width": "100px",
          }
        }
      >
        <goodwallet.svg />
      </div>
      <div
        className="css-view-1dbjc4n r-cursor-1loqt21 r-touchAction-1otgn73 r-transitionProperty-1i6wzkk r-userSelect-lrvibr"
        onClick={[Function]}
        onContextMenu={[Function]}
        onKeyDown={[Function]}
        style={
          Object {
            "WebkitAlignItems": "flex-end",
            "WebkitBoxAlign": "end",
            "WebkitTransitionDuration": "0s",
            "alignItems": "flex-end",
            "msFlexAlign": "end",
            "transitionDuration": "0s",
            "width": "37px",
          }
        }
        tabIndex="0"
      >
        <div
          className="css-text-901oao r-color-jwli3a"
          dir="auto"
          style={
            Object {
              "fontFamily": "gooddollar",
              "fontSize": "20px",
              "fontStyle": "normal",
              "fontWeight": "normal",
              "marginRight": "10px",
            }
          }
        >
          
        </div>
      </div>
    </div>
  </div>
</div>
`;
